intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java

```diff
@@ -0,0 +1,220 @@
+            final MessageBusConnection documentMessageBusConnection = project.getMessageBus().connect(disposable);
+
+            codeAnalyzerMessageBusConnection.subscribe(DaemonCodeAnalyzer.DAEMON_EVENT_TOPIC, new DaemonCodeAnalyzer.DaemonListenerAdapter() {
+                @Override
+                public void daemonFinished() {
+                    // We only receive global Code Analyzer events. This means that once such an event triggered,
+                    // we do not know for which file the analysis actually succeeded. Therefore we need to check
+                    // the so-called "dirty" state of the file for this particular analyzer to check whether
+                    // the analyzer actually finished. In this case, `null` indicates: the file is not dirty.
+                    if (analyzer.getFileStatusMap().getFileDirtyScope(intellijDocument, Pass.UPDATE_ALL) == null) {
+                        final MarkupModelImpl markupModel = (MarkupModelImpl) DocumentMarkupModel.forDocument(intellijDocument, project, true);
+
+                        markupModelListener.processWarningSnapshot(markupModel.getAllHighlighters());
+                        markupModel.addMarkupModelListener(disposable, markupModelListener);
+
+                        // We batch up changes and only transfer them on every save. This is in-line with the Eclipse
+                        // interface, which only exposes listeners for `POST_BUILD`. Therefore, cache all warnings in
+                        // {@link IntelliJMarkupModelListener#generatedWarnings} and {@link IntelliJMarkupModelListener#warnings}
+                        // and flush these warnings after the fact.
+                        documentMessageBusConnection.subscribe(AppTopics.FILE_DOCUMENT_SYNC, new FileDocumentManagerAdapter() {
+                            @Override
+                            public void beforeDocumentSaving(@NotNull com.intellij.openapi.editor.Document savedDocument) {
+                                if (intellijDocument.equals(savedDocument)) {
+                                    markupModelListener.flushForDocument();
+                                }
+                            }
+                        });
+                        codeAnalyzerMessageBusConnection.disconnect();
+                    }
+                }
+            });
+        });
+
+        return markupModelListener;
+    }
+
+    private void processWarningSnapshot(RangeHighlighter[] highlighters) {
+        if (highlighters.length == 0) {
+            return;
+        }
+
+        // markupModel.getAllHighlighters returns a list of RangeHighlighter,
+        // even though we know they are all instances of RangeHighlighterEx.
+        // Therefore do an instanceof check and then explicitly cast them
+        // to use them in the other methods
+        final Stream<RangeHighlighterEx> rangeHighlighters = Arrays.stream(highlighters)
```

Comment:
do these really need to be final?

Reply:
They are never changed nor should they, so this is mostly a hint to the compiler for optimization purposes.

Reply:
I would prefer "final" to be removed if it is not absolutely necessary to compile. It burdens people's mind as to why it can't just be a normal local variable.